### ISSUE TITLE
Navigation menu does no longer appear as submenu

### ISSUE THREAD

**Issue by fm-sys** [cgeo/cgeo#11216]
The navigation menu does no longer appear as submenu but as popup menu instead, which results in a way longer "move distance" for selecting the navigation action. 

![Screenshot_20210713_124716_cgeo geocaching](https://user-images.githubusercontent.com/64581222/125439498-fe706097-d665-48b6-ad22-5240cfb34d73.jpg)

**Comment by fm-sys** [comment]
No idea, but could it be related to b3fb279?

**Comment by moving-bits** [comment]
No idea... But from looking at the code I wonder why it ever has been displayed as sort of a menu - codewise it's a regular `AlertDialog`.

**Comment by fm-sys** [comment]
> Question is, though, whether it's better or worse this way? 
IMHO it's worse, as the distance between the action bar item and the item you want to choose is way longer. 



> No idea... But from looking at the code I wonder why it ever has been displayed as sort of a menu - codewise it's a regular `AlertDialog`.

AFAIK, the alert dialog is only a fallback. These are the interesting code places IIRC:

https://github.com/cgeo/cgeo/blob/0b0ee958afb729e17211b9ec713b25bf592e677f/main/src/cgeo/geocaching/CacheMenuHandler.java#L140-L147

https://github.com/cgeo/cgeo/blob/0c309bc5336659e65bcf160bc632a987754bad61/main/src/cgeo/geocaching/CacheDetailActivity.java#L646

**Comment by moving-bits** [comment]
Debugging into `CacheMenuHandler` lines 65ff:
```
} else if (menuItem == R.id.menu_navigate) {
    final NavigationSelectionActionProvider navigationProvider = (NavigationSelectionActionProvider) MenuItemCompat.getActionProvider(item);
    if (navigationProvider == null) {
        activityInterface.showNavigationMenu();
        return true;
    }
    return false;
}
```

I discovered that `getActionProvider()` always returns null, although the menu item definition in `cache_options.xml` defines an action provider: `cgeo.geocaching.apps.navi.NavigationSelectionActionProvider` (which is a valid class name of an existing class).
But why?

(BTW: How can you insert such nicely formatted source parts?)

**Comment by Lineflyer** [comment]
> (BTW: How can you insert such nicely formatted source **parts?)

Just use the URL to the source with a range of lines:
```
https://github.com/cgeo/cgeo/blob/0b0ee958afb729e17211b9ec713b25bf592e677f/main/src/cgeo/geocaching/CacheMenuHandler.java#L140-L147
```

**Comment by balu2003** [comment]
I'm not that good at java, but tell me why a menuitem object is created and used.
`} else if (menuItem == R.id.menu_navigate)`
BUT when calling ActionProvider, only item is used, not menuItem,
`MenuItemCompat.getActionProvider (item)`

**Comment by moving-bits** [comment]
`item` is the actual menu item object, whereas `menuItem` is just its id, see a few lines further up: `final int menuItem = item.getItemId();`

**Comment by balu2003** [comment]
ok, thank you, sorry for my stupid question:
I was a little bit confused, because in  CacheDetailActivity.java menuItem is created as MenuItem Object
`final MenuItem menuItem = menu.findItem(R.id.menu_default_navigation);
 ...
final NavigationActionProvider navAction = (NavigationActionProvider) MenuItemCompat.getActionProvider(menuItem); `

But in CacheMenuHandler.java, menuItem is used as int and as the id from the object Item (as MenuItem)
`final int menuItem = item.getItemId(); 
final NavigationSelectionActionProvider navigationProvider = (NavigationSelectionActionProvider) MenuItemCompat.getActionProvider(item); `

**Comment by moving-bits** [comment]
Yeah, it's not always consistent... Too much legacy code built up over the last ten years :-)

**Comment by moving-bits** [comment]
looks like `NavigationSelectionActionProvider` constructor needs `final Context context` as parameter to be recognized by the `actionProviderClass` mechanism (does not work with an `Activity`). Problem is, though, that then I won't have current activity at hand - which I need to set correct context for formatting the submenu...

**Comment by eddiemuc** [comment]
Can't you make the constructor accept a context parameter but pass an activity instance through it?

**Comment by eddiemuc** [comment]
Or, if this does not work for an reason (do the NaviApps really need an activity in `navigate` method, or can they use a context instead?), then pass a weak reference to the activity to method `NavigationSelectionActionProvider.initialize` and store it in the action (like the `Geocache` object passed as another parameter)

**Comment by moving-bits** [comment]
Thanks for the hints. The navigation apps would work with a `Context` object as well, had been doing so actually until app. two weeks ago. I can change the interface back to `Context`, so that it would be recognized by the `actionProviderClass` mechanism again. 

But then we are also back to the original problem of the navigation menu/popup formatting being broken #11147 :-(

**Comment by eddiemuc** [comment]
Might be fixed as well with PR #11397

**Comment by fm-sys** [comment]
> Might be fixed as well with PR #11397

Yes, does no longer happen. Thanks for the fix!
